File: packages/react-native-nitro-fetch/ios/MultipartBodyBuilder.swift
import Foundation

enum MultipartBodyBuilder {
  /// Builds an RFC 2046 multipart/form-data body and returns `(body, contentType)`.
  /// File URIs are read synchronously. `http(s)://` source URIs block the calling thread.
  static func build(parts: [NitroFormDataPart]) throws -> (Data, String) {
    let boundary = "NitroFetch-\(UUID().uuidString)"
    var body = Data()
    let crlf = "\r\n"

    for part in parts {
      body.append("--\(boundary)\(crlf)".data(using: .utf8)!)

      if let fileUri = part.fileUri {
        let fileName = part.fileName ?? "file"
        let mimeType = part.mimeType ?? "application/octet-stream"
        body.append("Content-Disposition: form-data; name=\"\(part.name)\"; filename=\"\(fileName)\"\(crlf)".data(using: .utf8)!)
        body.append("Content-Type: \(mimeType)\(crlf)\(crlf)".data(using: .utf8)!)

        let fileData = try readFileData(fileUri)
        body.append(fileData)
      } else {
        let value = part.value ?? ""
        body.append("Content-Disposition: form-data; name=\"\(part.name)\"\(crlf)\(crlf)".data(using: .utf8)!)
        body.append(value.data(using: .utf8)!)
      }

      body.append(crlf.data(using: .utf8)!)
    }

    body.append("--\(boundary)--\(crlf)".data(using: .utf8)!)
    return (body, "multipart/form-data; boundary=\(boundary)")
  }

  private static func readFileData(_ uri: String) throws -> Data {
    if uri.hasPrefix("http://") || uri.hasPrefix("https://") {
      guard let url = URL(string: uri) else {
        throw NSError(domain: "NitroFetch", code: -4, userInfo: [NSLocalizedDescriptionKey: "Invalid URL: \(uri)"])
      }
      return try Data(contentsOf: url)
    }
    let path = uri.hasPrefix("file://") ? String(uri.dropFirst(7)) : uri
    guard let data = FileManager.default.contents(atPath: path) else {
      throw NSError(domain: "NitroFetch", code: -4, userInfo: [NSLocalizedDescriptionKey: "Cannot read file at: \(uri)"])
    }
    return data
  }
}

File: packages/react-native-nitro-fetch/android/src/main/java/com/margelo/nitro/nitrofetch/MultipartBodyBuilder.kt
package com.margelo.nitro.nitrofetch

import android.net.Uri
import com.margelo.nitro.NitroModules
import java.io.ByteArrayOutputStream
import java.io.File
import java.util.UUID

object MultipartBodyBuilder {
  /// Builds an RFC 2046 multipart/form-data body and returns `(body, contentType)`.
  /// File URIs are read synchronously. `http(s)://` source URIs block the calling thread.
  fun build(parts: Array<NitroFormDataPart>): Pair<ByteArray, String> {
    val boundary = "NitroFetch-${UUID.randomUUID()}"
    val out = ByteArrayOutputStream()
    val crlf = "\r\n".toByteArray()

    for (part in parts) {
      out.write("--$boundary\r\n".toByteArray())

      val fileUri = part.fileUri
      if (fileUri != null) {
        val fileName = part.fileName ?: "file"
        val mimeType = part.mimeType ?: "application/octet-stream"
        out.write("Content-Disposition: form-data; name=\"${part.name}\"; filename=\"$fileName\"\r\n".toByteArray())
        out.write("Content-Type: $mimeType\r\n\r\n".toByteArray())

        val fileData = readFileBytes(fileUri)
        out.write(fileData)
      } else {
        val value = part.value ?: ""
        out.write("Content-Disposition: form-data; name=\"${part.name}\"\r\n\r\n".toByteArray())
        out.write(value.toByteArray(Charsets.UTF_8))
      }

      out.write(crlf)
    }

    out.write("--$boundary--\r\n".toByteArray())
    return Pair(out.toByteArray(), "multipart/form-data; boundary=$boundary")
  }

  private fun readFileBytes(uri: String): ByteArray {
    if (uri.startsWith("http://") || uri.startsWith("https://")) {
      val url = java.net.URL(uri)
      return url.openStream().use { it.readBytes() }
    }
    if (uri.startsWith("content://")) {
      val context = NitroModules.applicationContext
        ?: throw IllegalStateException("Cannot read content:// URI - no Android Context")
      val inputStream = context.contentResolver.openInputStream(Uri.parse(uri))
        ?: throw IllegalArgumentException("Cannot open content URI: $uri")
      return inputStream.use { it.readBytes() }
    }
    val path = if (uri.startsWith("file://")) uri.removePrefix("file://") else uri
    return File(path).readBytes()
  }
}
